calyx/src/frontend/workspace.rs

```diff
@@ -0,0 +1,204 @@
+use std::{
+    collections::HashSet,
+    fs,
+    path::{Path, PathBuf},
+};
+
+use super::{
+    ast::{ComponentDef, NamespaceDef},
+    parser,
+};
+use crate::{
+    errors::{CalyxResult, Error},
+    ir,
+};
+
+/// A Workspace represents all Calyx files transitively discovered while trying to compile a
+/// top-level file.
+///
+/// # Example
+/// When parsing a file `foo.futil`:
+/// ```text
+/// import "core.futil";
+///
+/// component main() -> () { ... }
+/// ```
+///
+/// The workspace gets the absolute path for `core.futil` and adds `main` to the set of defined
+/// components. `core.futil` is searched *both* relative to the current file and the library path.
+/// Next `core.futil` is parsed:
```

Comment:
What happens if there's a local `core.futil`? Does it register that as a conflict?

Reply:
No. It'll just prioritize one of the two. Not sure if that's the right thing to do or not.